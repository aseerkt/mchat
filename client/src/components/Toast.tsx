import { VariantProps, cva } from 'cva'
import { useEffect } from 'react'
import { createPortal } from 'react-dom'

const toastVariants = cva(
  'fixed flex text-white justify-between gap-4 md:min-w-[250px] shadow-lg border rounded p-3 bottom-6 z-10 transition-all',
  {
    variants: {
      severity: {
        error: 'bg-red-800',
        info: 'bg-blue-800',
        success: 'bg-green-800',
      },
      state: {
        open: 'right-6',
        close: '-right-full',
      },
    },
    defaultVariants: {
      state: 'close',
      severity: 'info',
    },
  },
)

export type ToastVariantProps = Omit<
  VariantProps<typeof toastVariants>,
  'state'
>

type ToastProps = ToastVariantProps & {
  open: boolean
  onOpenChange: (open: boolean) => void
  children: React.ReactNode
}

export const Toast: React.FC<ToastProps> = ({
  open,
  onOpenChange,
  severity,
  children,
}) => {
  const handleClose = () => onOpenChange(false)

  useEffect(() => {
    if (open) {
      const timeout = setTimeout(handleClose, 3000)
      return () => {
        clearTimeout(timeout)
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [open])

  return createPortal(
    <div
      className={toastVariants({ severity, state: open ? 'open' : 'close' })}
    >
      <>{children}</>
      <button
        aria-label='close toast'
        className='inline-flex h-5 w-5 items-center justify-center rounded-full text-white hover:bg-gray-100'
        onClick={handleClose}
      >
        x
      </button>
    </div>,
    document.body,
  )
}
