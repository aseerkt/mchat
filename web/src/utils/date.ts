export const isToday = (date: Date) => {
  const now = new Date()

  return (
    date.getDate() === now.getDate() &&
    date.getMonth() === now.getMonth() &&
    date.getFullYear() === now.getFullYear()
  )
}

export const isYesterday = (date: Date) => {
  const now = new Date()
  const yesterday = new Date(now)
  yesterday.setDate(now.getDate() - 1)
  return (
    date.getDate() === yesterday.getDate() &&
    date.getMonth() === yesterday.getMonth() &&
    date.getFullYear() === yesterday.getFullYear()
  )
}
export const formateChatDate = (date: string | number | Date) => {
  const dateObj =
    typeof date === 'number' || typeof date === 'string' ? new Date(date) : date

  if (isToday(dateObj)) {
    return (
      'Today ' +
      dateObj.toLocaleTimeString([], {
        hour: '2-digit',
        minute: '2-digit',
      })
    ) // e.g., "14:23"
  } else if (isYesterday(dateObj)) {
    return (
      'Yesterday ' +
      dateObj.toLocaleTimeString([], {
        hour: '2-digit',
        minute: '2-digit',
      })
    )
  } else {
    return dateObj.toTimeString() // e.g., "June 18, 2023"
  }
}

export const formatGroupDate = (date: string | number | Date) => {
  const dateObj =
    typeof date === 'number' || typeof date === 'string' ? new Date(date) : date
  const today = new Date()
  const yesterday = new Date(today)
  yesterday.setDate(yesterday.getDate() - 1)

  const isSameDay =
    dateObj.getDate() === today.getDate() &&
    dateObj.getMonth() === today.getMonth() &&
    dateObj.getFullYear() === today.getFullYear()

  const isSameYesterday =
    dateObj.getDate() === yesterday.getDate() &&
    dateObj.getMonth() === yesterday.getMonth() &&
    dateObj.getFullYear() === yesterday.getFullYear()

  if (isSameDay) {
    // Format time if the date is today
    return `${('0' + dateObj.getHours()).slice(-2)}:${('0' + dateObj.getMinutes()).slice(-2)}`
  } else if (isSameYesterday) {
    // Show "Yesterday" if the date is yesterday
    return 'Yesterday'
  } else if (
    dateObj >= new Date(today.setDate(today.getDate() - today.getDay()))
  ) {
    // Show day of the week for dates within this week
    const days = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat']
    return days[dateObj.getDay()]
  } else {
    // Show date in MMM DD format for older dates
    const options = { month: 'short', day: 'numeric' } as const
    return dateObj.toLocaleDateString('en-US', options)
  }
}

// Example usage:
