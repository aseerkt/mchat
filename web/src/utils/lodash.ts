export const debounce = <TFnArgs extends unknown[], TFnResult>(
  fn: (...args: TFnArgs) => TFnResult,
  ms = 300,
): ((...args: TFnArgs) => void) => {
  let timeout: NodeJS.Timeout | null = null

  return function (...args: TFnArgs) {
    if (timeout) {
      clearTimeout(timeout)
    }

    timeout = setTimeout(() => {
      fn(...args)
      timeout = null
    }, ms)
  }
}

export const memoize = <TFnArgs extends unknown[], TFnResult>(
  fn: (...args: TFnArgs) => TFnResult,
) => {
  const cache = new Map<string, TFnResult>()

  return (...args: TFnArgs) => {
    const key = JSON.stringify(args)

    if (cache.has(key)) {
      return cache.get(key)!
    }
    const result = fn(...args)
    cache.set(key, result)
    return result
  }
}
