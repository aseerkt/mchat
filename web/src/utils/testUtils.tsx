export const timeout = (timeout: number) =>
  new Promise(res => setTimeout(res, timeout))
