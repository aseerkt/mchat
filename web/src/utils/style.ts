import { clsx, type ClassValue } from 'clsx'
import { twMerge } from 'tailwind-merge'
import { memoize } from './lodash'

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs))
}

export const stringToColor = memoize((str: string) => {
  let hash = 0
  for (let i = 0; i < str.length; i++) {
    hash = str.charCodeAt(i) + ((hash << 5) - hash)
  }

  hash = Math.abs(hash)

  const r = (hash & 0xff0000) >> 16
  const g = (hash & 0x00ff00) >> 8
  const b = hash & 0x0000ff

  const bgColor = `#${((1 << 24) + (r << 16) + (g << 8) + b).toString(16).slice(1).toUpperCase()}`

  // Calculate luminance to determine appropriate text color
  const luminance = (0.299 * r + 0.587 * g + 0.114 * b) / 255
  const textColor = luminance > 0.5 ? '#000000' : '#FFFFFF'

  return { bgColor, textColor }
})
