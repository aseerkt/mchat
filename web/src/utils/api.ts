import { config } from '../config'
import { getToken } from './token'

export const stringifyQueryParams = <
  TQueryParams extends Record<string, unknown>,
>(
  params: TQueryParams,
) => {
  const queryParams = new URLSearchParams()
  Object.keys(params).forEach(key => {
    const value = params[key] ? String(params[key]) : undefined
    if (value) {
      queryParams.append(key, value)
    }
  })
  return queryParams.toString()
}

export const fetcher = async (path: string, options?: RequestInit) => {
  const token = getToken()
  const res = await fetch(`${config.backendUrl}/api/${path}`, {
    ...options,
    headers: {
      'Content-Type': 'application/json',
      Authorization: token ? `Bearer ${token}` : '',
      ...options?.headers,
    },
  })

  const isJsonType = res.headers
    .get('Content-Type')
    ?.includes('application/json')

  const data = isJsonType ? await res.json() : await res.text()

  if (res.ok) {
    return data
  } else {
    throw new Error(data.message)
  }
}
