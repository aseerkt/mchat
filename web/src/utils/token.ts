import { fetcher } from './api'

class Token {
  #token?: string

  get() {
    return this.#token
  }

  set(token: string) {
    this.#token = token
  }

  remove() {
    this.#token = undefined
  }

  async recreate() {
    const { accessToken } = await fetcher('refresh')
    this.#token = accessToken
  }

  decode() {
    if (!this.#token) return

    const base64Url = this.#token.split('.')[1]
    const base64 = base64Url.replace(/-/g, '+').replace(/_/g, '/')
    const jsonPayload = decodeURIComponent(
      window
        .atob(base64)
        .split('')
        .map(function (c) {
          return '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2)
        })
        .join(''),
    )
    return JSON.parse(jsonPayload)
  }
}

export const accessToken = new Token()
