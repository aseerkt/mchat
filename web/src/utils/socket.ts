import { TypedSocket } from '@/interfaces/socket.interface'
import { io } from 'socket.io-client'
import { config } from '../config'
import { toast } from '../hooks/useToast'
import { accessToken } from './token'

let socket: TypedSocket

export const getSocketIO = () => {
  if (!socket) {
    socket = io(config.backendUrl, {
      reconnectionAttempts: 5,
      autoConnect: false,
      auth(cb) {
        const token = accessToken.get()

        if (!token) {
          toast({ title: 'socket: no auth token', severity: 'error' })
        }
        cb({ token })
      },
    })
  }
  return socket
}
