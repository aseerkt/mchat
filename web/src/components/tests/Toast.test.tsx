import { timeout } from '@/utils/testUtils'
import { render } from '@testing-library/react'
import { Toast } from '../Toast'

describe('<Toast / > tests', () => {
  it('should show toast only for 3 sec', async () => {
    const onOnOpenSpy = vi.fn()
    const { baseElement } = render(
      <Toast open={true} onOpenChange={onOnOpenSpy}>
        This is a toast!
      </Toast>,
    )
    expect(baseElement).toMatchSnapshot('toast visible')
    await timeout(3500)
    expect(onOnOpenSpy).toBeCalledWith(false)
  })
})
