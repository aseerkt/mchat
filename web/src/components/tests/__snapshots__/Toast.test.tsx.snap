// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Toast / > tests > should show toast only for 3 sec > toast hidden 1`] = `
<body>
  <div />
  <div
    class="fixed flex text-white justify-between gap-4 md:min-w-[250px] shadow-lg border rounded p-3 bottom-6 z-10 transition-all bg-blue-800 right-6"
  >
    This is a toast!
    <button
      aria-label="close toast"
      class="inline-flex h-5 w-5 items-center justify-center rounded-full text-white hover:bg-gray-100"
    >
      x
    </button>
  </div>
</body>
`;

exports[`<Toast / > tests > should show toast only for 3 sec > toast visible 1`] = `
<body>
  <div />
  <div
    class="fixed flex text-white justify-between gap-4 md:min-w-[250px] shadow-lg border rounded p-3 bottom-6 z-10 transition-all bg-blue-800 right-6"
  >
    This is a toast!
    <button
      aria-label="close toast"
      class="inline-flex h-5 w-5 items-center justify-center rounded-full text-white hover:bg-gray-100"
    >
      x
    </button>
  </div>
</body>
`;

exports[`<Toast / > tests > should show toast only for 3 sec 1`] = `
<body>
  <div />
  <div
    class="fixed flex text-white justify-between gap-4 md:min-w-[250px] shadow-lg border rounded p-3 bottom-6 z-10 transition-all bg-blue-800 right-6"
  >
    This is a toast!
    <button
      aria-label="close toast"
      class="inline-flex h-5 w-5 items-center justify-center rounded-full text-white hover:bg-gray-100"
    >
      x
    </button>
  </div>
</body>
`;
